### PR TITLE
chore: update ink to v3

### LINKED ISSUE #19658: Ink CLI logger truncates long messages
## Description

Some of our warnings are quite long. If warning text doesn't fit into terminal window width, `ink` will wrap the text. But when this happens it will also truncate the last line of the text.

As a result, some warnings are truncated which causes confusion for people seeing them. Example: https://github.com/gatsbyjs/gatsby/issues/19314

This should be probably fixed upstream - https://github.com/vadimdemedes/ink/issues/245

But then we should update as soon as it's fixed. Other options to fix this are:

1. Do not use `textWrap="wrap"` for messages
2. Wrap a message text with a separate box and move `textWrap` to this box (but this will change console output a bit)

### PR DESCRIPTION
**Pull request by herecydev**
## Description

Updates [ink](https://github.com/vadimdemedes/ink) to v3. Pretty straightforward migration for most things; but some small bugs to fix:

- [x] Test gatsby-recipes
- [x] Duplicate "Done building" log
- [x] Empty line after all messages (can replicate with ["foo, "bar"] as well, so possibly an ink issue)
- [x] Fix spinner, related: https://github.com/vadimdemedes/ink-spinner/issues/10
- [x] Fix tests

image

Closes #19658

### CODE REVIEW

**Comment by vladar** [comment]
I converted the PR to draft until it's ready. Thanks for doing this upgrade! 💜

**Comment by herecydev** [comment]
Still looking into this duplicate message at the end; I've created a simplified `cli.tsx` that demonstrates the weirdness. There are a total of 27 messages that need rendering in the `Static` ink component. If I only render 26 it's fine (so it's not a last message problem). If I add additional items to the list they are all duplicated

image

Outputs the following:

image

It seems consistently after the 26th index it starts repeating the 26th, 27th. 28th, etc messages.

**Comment by sidharthachatterjee** [comment]
@herecydev Thanks for picking this up!

**Comment by herecydev** [comment]
No problem, I'm very close to getting an understanding of the duplicate log (tracking that in main ink repo). Ink-spinner still needs fixing as well; otherwise this shouldn't be too far away

**Comment by hoobdeebla** [comment]
It looks like the integration test is trying to call `del-cli`, but the [del-cli docs](/sindresorhus/del-cli) indicate that the command for doing so is just `del` on Unix-like systems. Try editing that line like so:

```diff
-   const clean = dir => execa(`yarn`, ["del-cli", dir])`
+   const clean = dir => execa(`yarn`, ["del", dir])`
```

Hopefully this helps :)

**Comment by herecydev** [comment]
I'm on windows and it explicity calls out using `del-cli` there:

> Windows users: Since $ del is already a builtin command on Windows, you need to use $ del-cli there.

**Comment by hoobdeebla** [comment]
Gotcha. I cloned your PR just now on Windows and WSL and the `new` test worked in both places as-is for me, and it worked in the last CI build as-is too. Not sure I can offer any further insight here but happy to help if needed

**Comment by herecydev** [comment]
I'll give it another try then 🤔 the errors look fairly straightforward to fix regardless

**Comment by herecydev** [comment]
image

Still struggling to run these integration tests, they are hanging indefinitely

**Comment by KyleAMathews** [comment]
Was trying to test this locally just now with `gatsby-dev`. It kept failing with a React error saying it received an object instead of a component.

**Comment by KyleAMathews** [comment]
Not sure if that was my local setup or something not being right or what.

**Comment by hoobdeebla** [comment]
> Was trying to test this locally just now with `gatsby-dev`. It kept failing with a React error saying it received an object instead of a component.

I got the same error when testing locally - the PR updates ink and ink-spinner as production dependencies rather than devDeps. This was changed upstream with #27058. To fix it, I removed ink and ink-spinner from deps then updated them in devDeps. All CLI tests passed, and `gatsby-dev` worked without erroring after this change.

**Comment by herecydev** [comment]
@hoobdeebla feel free to create a PR into my branch and I'll merge it! 👍

**Comment by hoobdeebla** [comment]
This should be ready @wardpeet. I don't have direct write access to this PR, so I can't fix the 2 conflicts.

**Comment by pieh** [comment]
> This should be ready @wardpeet. I don't have direct write access to this PR, so I can't fix the 2 conflicts.

I'll fix the conflicts and do some more testing ;)

**Comment by pieh** [comment]
> Was trying to test this locally just now with gatsby-dev. It kept failing with a React error saying it received an object instead of a component.

> I got the same error when testing locally - the PR updates ink and ink-spinner as production dependencies rather than devDeps. This was changed upstream with #27058. To fix it, I removed ink and ink-spinner from deps then updated them in devDeps. All local CLI tests passed, and gatsby-dev worked without erroring after this change.

So interesting thing I discovered as I worked on resolving conflicts on this PR + getting it through finish line.

https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-cli/src/reporter/start-logger.ts#L12-L16 checks if `ink` is installed - and switches to legacy logger if it's not. #27058 in practice broke auto-selecting ink for majority of users. I didn't see this problem because I was testing it inside `gatsbyjs/gatsby` monorepo which has ink installed, but user sites actually don't install `ink` package anymore ...

After fixing above check (well removing it, because it's not needed anymore) I also did hit this problem and this is because 
```
import InkSpinner from "ink-spinner"
console.log({ InkSpinner })
{ InkSpinner: { default: [Function: Spinner] } }
```

So rollup (well, our setup at least) doesn't "prefer" default so instead of `InkSpinner` being just component `InkSpinner` here actually is "namespace". I will use this trick https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-recipes/src/cli/index.js#L344 to workaround it for now to not mess with bundling setup (and avoid prolonging this PR any more).

I also want to sure couple more niceties that this upgrade - ink@2 (or maybe the way we use it now) sometimes truncates longer messages ( https://github.com/gatsbyjs/gatsby/issues/19658 ) - this PR (so this might be either ink upgrade or component changes done here or both) actually seems to fix that - see https://www.loom.com/share/925e12ce153c49d2a0ddb7d1ae7bb1a6 

- left `yurnalist` logger - problem - multiple "self updating" things are flickering and "fighting" to control terminal
- center - current ink@2 - messages get truncated (they are supposed to have 5 paragraphs and end with clear `-- END --`
- right - this PR with couple of adjustments that I will push shortly - no flickering spinners/progress bars and no truncated messages - clear winner

File: packages/gatsby-cli/src/reporter/loggers/ink/context.tsx [hunk 1]
```diff
@@ -17,7 +17,7 @@ export const StoreStateProvider: React.FC = ({
     (getStore().getState() as any) as IGatsbyState
   )
 
-  useEffect(
+  useLayoutEffect(
```
Comment (pieh) [inline packages/gatsby-cli/src/reporter/loggers/ink/context.tsx:20]: This seems needed for quick commands (like `gatsby options`) that are in practice short sync ink@3 doesn't seem to run `useEffect` before we start printing things and result in no output in those cases

**Comment by pieh** [comment]
I published `gatsby@ink3` / `gatsby-cli@ink3` if anyone want to do some checking on how it works with current state of things - I will be doing some more cheks too

**Comment by pieh** [comment]
Ok, it seems like I did broke `gatsby recipes` in most recent changes:
```
➜  gatsby-ink3 git:(master) ✗ gatsby recipes
Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at hd (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:174:108)
    at ld (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:59:104)
    at Q (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:81:38)
    at ig (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:161:347)
    at hg (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:139:22)
    at bg (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:138:366)
    at c (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:132:214)
    at b (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:30:115)
    at Mb (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/scheduler/cjs/scheduler.production.min.js:18:437)
    at hc (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.production.min.js:29:325)

  ERROR Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message
       or use the non-minified dev environment for full errors and additional helpful warnings.

 /Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.produc
 tion.min.js:174:108

 -hd (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.p
    roduction.min.js:174:108)
 -ld (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.p
    roduction.min.js:59:104)
 -Q (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.pr
   oduction.min.js:81:38)
 -ig (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.p
    roduction.min.js:161:347)
 -hg (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.p
    roduction.min.js:139:22)
 -bg (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.p
    roduction.min.js:138:366)
 -c (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.pr
   oduction.min.js:132:214)
 -b (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.pr
   oduction.min.js:30:115)
 - Mb (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/scheduler/cjs/scheduler.production.min.js:18:437)
 -hc (/Users/misiek/.nvm/versions/node/v12.18.2/lib/node_modules/gatsby-cli/node_modules/ink/node_modules/react-reconciler/cjs/react-reconciler.p
    roduction.min.js:29:325)
```

**Comment by pieh** [comment]
New version were published for `gatsby@ink3` / `gatsby-cli@ink3` that seem to fix recipes (again)

**Review by wardpeet — COMMENTED** [review]
One small nit

File: packages/gatsby-cli/src/reporter/loggers/ink/components/develop.tsx [hunk 2]
```diff
@@ -24,7 +24,12 @@ const getLabel = (
 
 // Track the width and height of the terminal. Responsive app design baby!
 const useTerminalResize = (): Array<number> => {
-  const { stdout } = useContext(StdoutContext)
+  const { stdout } = useStdout()
+
+  if (!stdout) {
+    return [0]
```
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby-cli/src/reporter/loggers/ink/components/develop.tsx:30]: This is to keep TS gods happy - `stdout` from `useStdout` have type definition: `readonly stdout?: NodeJS.WriteStream;` ( https://unpkg.com/browse/ink@3.0.8/build/components/StdoutContext.d.ts )

**Review by wardpeet — APPROVED** [review]

**Pull request merged**